test/e2e/kappcontroller/imgpkg_bundle_test.go

```diff
@@ -123,8 +118,226 @@ spec:
+      - kapp: {}
+---
+apiVersion: packaging.carvel.dev/v1alpha1
+kind: PackageInstall
+metadata:
+  name: %[2]s
+  namespace: %[1]s
+  annotations:
+    kapp.k14s.io/change-group: kappctrl-e2e.k14s.io/packageinstalls
+    kapp.k14s.io/change-rule: "upsert after upserting package"
+spec:
+  syncPeriod: 30s
+  serviceAccountName: kappctrl-e2e-ns-sa
+  packageRef:
+    refName: pkg.test.carvel.dev
+    versionSelection:
+      constraints: 1.0.0
+`, env.Namespace, pkgiName, registryNamespace) + sas.ForNamespaceYAML()
+
+		kapp.RunWithOpts([]string{"deploy", "-a", pkgiName, "-f", "-"}, e2e.RunOpts{
+			StdinReader:  strings.NewReader(pkgiYaml),
+			OnErrKubectl: []string{"get", "app", pkgiName, "-oyaml"},
+		})
+
+		kubectl.Run([]string{"get", "configmap", "e2e-test-map"})
+		output := kubectl.Run([]string{"get", "app", pkgiName, "-oyaml"})
+		var appcr v1alpha1.App
+		require.NoError(t, yaml.Unmarshal([]byte(output), &appcr))
+		require.Equal(t, 0, appcr.Status.Fetch.ExitCode)
+		initialFetchUpdate = appcr.Status.Fetch.UpdatedAt
+	})
+	defer kapp.Run([]string{"delete", "-a", pkgiName})
+
+	logger.Section("remove registry", func() {
+		kapp.Run([]string{"delete", "-a", registryName, "-n", registryNamespace})
+		isRegistryPresent = false
+	})
+
+	logger.Section("wait for reconciliation and ensure no error occur while fetching", func() {
+		waitForPKGIReconciliationOrFail(t, kubectl, pkgiName, initialFetchUpdate)
+
+		var appcr v1alpha1.App
+		output := kubectl.Run([]string{"get", "app", pkgiName, "-oyaml"})
+		require.NoError(t, yaml.Unmarshal([]byte(output), &appcr))
+		require.Equal(t, 0, appcr.Status.Fetch.ExitCode)
+	})
+}
+
+func Test_PackageRepo_FetchFromCaching(t *testing.T) {
+	env := e2e.BuildEnv(t)
+	logger := e2e.Logger{}
+	kapp := e2e.Kapp{t, env.Namespace, logger}
+	kubectl := e2e.Kubectl{t, env.Namespace, logger}
+	sas := e2e.ServiceAccounts{env.Namespace}
+	isRegistryPresent := false
+
+	// If this changes, the skip-tls-verify domain must be updated to match
+	pkgrName := "pkgr-to-test-caching"
+	registryNamespace := "registry"
+	registryName := "test-registry"
+
+	cleanUp := func() {
+		// Delete App with kubectl first since kapp
+		// deletes ServiceAccount before App
+		kapp.Run([]string{"delete", "-a", pkgrName})
+		kapp.Run([]string{"delete", "-a", "secret-export"})
+		if isRegistryPresent {
+			kapp.Run([]string{"delete", "-a", registryName, "-n", registryNamespace})
+		}
+	}
+	cleanUp()
+	defer cleanUp()
+
+	logger.Section("deploy registry without authentication", func() {
+		kapp.Run([]string{
+			"deploy", "-a", registryName,
+			"-f", "../assets/registry/no-auth-registry.yml",
+			"-f", "../assets/registry/registry-contents.yml",
+		})
+	})
+
+	var initialFetchUpdate metav1.Time
+
+	logger.Section("deploy PackageRepository that auths to registry", func() {
+		pkgrYaml := fmt.Sprintf(`---
+apiVersion: packaging.carvel.dev/v1alpha1
+kind: PackageRepository
+metadata:
+  name: %[2]s
+  namespace: %[1]s
+spec:
+  syncPeriod: 30s
```

Comment:
instead of 30s delay here, can we set this to be higher, and do an explicit kick in the test? (would in theory be more reliable and faster)

Reply:
I tried the pause/unpause trick but I was not getting a reconciliation. I can try to do it again to see if it started working now

Reply:
lmk if you need second set of eyes.